### PR TITLE
Fix terminating containers

### PR DESCRIPTION
**Pull request by l0rd**
Containers components should be non-terminating (c.f. [point n.1 here](https://github.com/devfile/api/issues/679)). 

This PR adds a script to check that containers components are non terminating (`./tests/check_non_terminating.sh`, used as PR check too):

[asciicast](https://asciinema.org/a/a91erkXlTrFhqobkFdY6v2pvp)

image

It also patches the registry devfiles that failed to pass the test with:

```diff
container:
   image: (...)
+  args: ["tail", "-f", "/dev/null"]
```

or

```diff
container:
   image: (...)
+  command: ["tail", "-f", "/dev/null"]
```

This is the list of patched devfiles:

```
stacks/dotnet50/devfile.yaml
stacks/dotnet60/devfile.yaml
stacks/dotnetcore31/devfile.yaml
stacks/java-openliberty-gradle/devfile.yaml
stacks/java-openliberty/devfile.yaml
stacks/java-quarkus/devfile.yaml
stacks/java-websphereliberty-gradle/devfile.yaml
stacks/java-websphereliberty/devfile.yaml
stacks/java-wildfly-bootable-jar/devfile.yaml
stacks/nodejs-angular/devfile.yaml
stacks/nodejs-nextjs/devfile.yaml
stacks/nodejs-nuxtjs/devfile.yaml
stacks/nodejs-react/devfile.yaml
stacks/nodejs-svelte/devfile.yaml
stacks/nodejs-vue/devfile.yaml
stacks/nodejs/devfile.yaml
stacks/php-laravel/devfile.yaml
stacks/python-django/devfile.yaml
stacks/python/devfile.yaml
```

### Which issue(s) this PR fixes:

https://github.com/devfile/api/issues/681

### PR acceptance criteria:

- [x] Contributing guide
- [x] Test automation
- [ ] Documentation

### How to test changes / Special notes to the reviewer:

I have added the following lines in the repo README file **Running the tests** section:

> 3) From the root of this repository, run `tests/check_non_terminating.sh`. 
>   
>     - The test script will validate each devfile stack under `stacks/`, verifying that the components of type container are terminating. 
>        - The test script retrieves the `image`, `command` and `args` of a container component and uses them to run a pod and wait until it reaches the `Running` state:
>           ```bash
>           kubectl run test-terminating -n default --attach=false --restart=Never --image="<image>" --command=true -- "<command>" "<args>"
>           ```
>     - Each container component **must** be non-terminating. If the default `image` entrypoint is terminating an `args` (preferred) or `command` should be specified in the defile (e.g. `["tail", "-f", "/dev/null"]`).

### CODE REVIEW

**Comment by johnmcollier** [comment]
It looks like the changes aren't playing nicely with odo:
```
 ✗  Failed to start component with name "java-openliberty-gradle". Error: Failed to create the component: the supervisord program devrun not found
+ error=true
```

All of the stacks are failing with the same error

@l0rd @kadel

File: stacks/dotnet50/devfile.yaml
Comment (amisevsk) [reply stacks/dotnet50/devfile.yaml:28]: This is also widely used in the old Che devfile v1 images, e.g. https://github.com/eclipse-che/che-devfile-registry/blob/main/dockerfiles/entrypoint.sh

Comment (l0rd) [reply stacks/dotnet50/devfile.yaml:28]: @amisevsk thanks I will test that.

Comment (l0rd) [reply stacks/dotnet50/devfile.yaml:28]: After using `args` instead of `command` these stacks are still failing:

```
nodejs-nextjs nodejs java-quarkus php-laravel go nodejs-nuxtjs nodejs-vue python java-websphereliberty dotnet50 nodejs-react dotnet60 nodejs-svelte java-websphereliberty-gradle java-openliberty-gradle java-openliberty nodejs-angular dotnetcore31
```

Comment (amisevsk) [reply stacks/dotnet50/devfile.yaml:28]: Here's a summary of the entrypoint for each image:

|STACK NAME                    |IMAGE                                                           |ENTRYPOINT|
|-|-|-|
|nodejs-nextjs                 |node:lts-slim                                                   |["docker-entrypoint.sh"]|
|nodejs                        |registry.access.redhat.com/ubi8/nodejs-14:latest                |["container-entrypoint"]|
|java-quarkus                  |registry.access.redhat.com/ubi8/openjdk-11                      |null|
|php-laravel                   |composer:2.1.11                                                 |["/docker-entrypoint.sh"]|
|go                            |golang:latest                                                   |null|
|nodejs-nuxtjs                 |node:lts                                                        |["docker-entrypoint.sh"]|
|nodejs-vue                    |node:lts-slim                                                   |["docker-entrypoint.sh"]|
|python                        |quay.io/eclipse/che-python-3.7:nightly                          |["/entrypoint.sh"]|
|java-websphereliberty         |icr.io/appcafe/websphere-liberty-devfile-stack:22.0.0.1         |null|
|dotnet50                      |registry.access.redhat.com/ubi8/dotnet-50:5.0                   |["container-entrypoint"]|
|nodejs-react                  |node:lts-slim                                                   |["docker-entrypoint.sh"]|
|dotnet60                      |registry.access.redhat.com/ubi8/dotnet-60:6.0                   |["container-entrypoint"]|
|nodejs-svelte                 |node:lts-slim                                                   |["docker-entrypoint.sh"]|
|java-websphereliberty-gradle  |icr.io/appcafe/websphere-liberty-devfile-stack:22.0.0.1-gradle  |null|
|java-openliberty-gradle       |icr.io/appcafe/open-liberty-devfile-stack:22.0.0.1-gradle       |null|
|java-openliberty              |icr.io/appcafe/open-liberty-devfile-stack:22.0.0.1              |null|
|nodejs-angular                |node:lts-slim                                                   |["docker-entrypoint.sh"]|
|dotnetcore31                  |registry.access.redhat.com/ubi8/dotnet-31:3.1                   |["container-entrypoint"]|

Checking a few of these, it seems they should be fine if `args` is set to `["tail", "-f", "/dev/null"]` and `command` is left unchanged (e.g. the node:lts and node:lts-slim images use an entrypoint that should just `exec` the args) -- running `docker run -it --rm $image tail -f /dev/null` does result in a non-terminating container.

I don't know if `odo` is doing anything fancy for these images, but if it was always just overriding the command/args then it _should_ work as far as I can tell.

**Comment by johnmcollier** [comment]
CI now passing with odo v2.5.1 👍

**Comment by johnmcollier** [comment]
@l0rd Can this be moved out of Draft / WIP now?

**Comment by kadel** [comment]
> @l0rd Can this be moved out of Draft / WIP now?

no, please don't merge this yet. 

The majority of users still use 2.5.0. and more importantly, the OpenShift connector (vscode plugin) is still based on 2.5.0. If you merge this now, it will break them.

**Comment by l0rd** [comment]
@kadel I will wait for your approval to merge this

**Comment by l0rd** [comment]
@kadel can we merge this one now?

**Comment by kadel** [comment]
> @kadel can we merge this one now?

We are getting close. Currently, we a little over 60% of users on a version that includes fix.
Let me just double-check with IDE plugins if they have the majority of their users on the fixed version as well.
BTW: I will be on PTO for two weeks. If it is possible let's wait with this till I'm back. It will break all odo < 2.5.1 versions and I would like to be here for that.

**Comment by kadel** [comment]
The latest versions of all IDE plugins based on odo are on 2.5.1.
The majority of odo users that have telemetry enabled are on 2.5.1 as well.

🟢 We can merge this PR.

**Comment by l0rd** [comment]
🎉 Thank you @kadel I will rebase and merge the PR Tuesday when I come back from PTO.

**Comment by l0rd** [comment]
🎉 Thank you @kadel I will rebase and merge the PR Tuesday when I come back from PTO.

**Comment by johnmcollier** [comment]
@l0rd will you be rebasing soon?

**Comment by l0rd** [comment]
@johnmcollier I am still in vacation. I will be back next week and will rebase it.

**Comment by l0rd** [comment]
@johnmcollier do you know why the `ci/prow/v4.7-registry-test` PR check is failing?

**Comment by kadel** [comment]
this this change using odo v3 tests that I'm currently working on, and all tests that were passing before this change are passing with this as well. 
/lgtm

**Comment by elsony** [comment]
It looks like the test failure is due to the recent issues with the OpenShift CI

File: tests/check_non_terminating.sh
Comment (amisevsk) [inline tests/check_non_terminating.sh:86]: Instead of using `2>/dev/null 1>/dev/null` it might be slightly cleaner to append `> /dev/null 2>&1`.
```suggestion
        kubectl run test-terminating -n default --attach=false --restart=Never --image="$_int_image" > /dev/null 2>&1
```

File: tests/check_non_terminating.sh
Comment (amisevsk) [inline tests/check_non_terminating.sh:98]: The run commands hard-code the namespace (`default`) whereas the variable is used here.

Comment (amisevsk) [inline tests/check_non_terminating.sh:98]: There's some fragility in this check -- a pod may briefly enter the Running state even if it is a terminating container, if e.g. it runs some entrypoint that does not resolve immediately. However, for now this likely sufficient.

I'm able to trick this test (most of the time, at least, on OpenShift) with
```bash
kubectl run test-terminating \
  --attach=false \
  --restart=Never \
  --image="registry.access.redhat.com/ubi8/nodejs-16:latest" \
  -- sleep 0.5s \
  && kubectl get po -w
```
but this could naturally arise if the pod attempts to start a server or similar.

For the future, perhaps a safer check would be to e.g. start a pod for each image in devfiles in parallel, then wait some period of time and verify that the pods still have the Ready condition.

File: tests/check_non_terminating.sh
Comment (amisevsk) [inline tests/check_non_terminating.sh:114]: Not sure if this is left over from an earlier refactor -- should this be something like

```suggestion
YQ_PATH=${YQ_PATH:-yq}
```

Comment (amisevsk) [reply tests/check_non_terminating.sh:114]: Also this would be very useful to have as this script uses `mikefarah/yq` whereas my default installed binary is the `kislyuk/yq` version. To use the script I would need to `export YQ_PATH=...`.

File: tests/check_non_terminating.sh
Comment (amisevsk) [inline tests/check_non_terminating.sh:100]: Using the `--force` argument, I get the warning
> warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.

If the attempt is to make the script run more quickly, perhaps naming the containers after the image being tested and passing `--wait=false` would be safer.

File: tests/check_non_terminating.sh
Comment (amisevsk) [inline tests/check_non_terminating.sh:79]: Creating a namespace specifically for the test might make cleanup easier, as you could just `oc delete project test-terminating` to make sure everything is cleaned up. The `default` namespace can have unexpected restrictions that are not present in non-default namespaces, and would prevent someone from running this test in a cluster where they are not cluster-admin.

**Review by amisevsk — APPROVED** [review]
Tested the script and it works as expected. I've left a number of optional comments, feel free to ignore/defer to a different PR.

File: tests/check_non_terminating.sh
Comment (l0rd) [reply tests/check_non_terminating.sh:86]: Makes sense, I will update that.

File: tests/check_non_terminating.sh
Comment (l0rd) [reply tests/check_non_terminating.sh:98]: Thanks, I will fix that

File: tests/check_non_terminating.sh
Comment (l0rd) [reply tests/check_non_terminating.sh:114]: 👍

File: tests/check_non_terminating.sh
Comment (l0rd) [reply tests/check_non_terminating.sh:100]: You are right, I will remove `--force`

**Comment by l0rd** [comment]
@amisevsk thanks for the review, I will address the comments later today.

**Comment by johnmcollier** [comment]
> @johnmcollier do you know why the ci/prow/v4.7-registry-test PR check is failing?

Sorry, just saw this. Not sure what's going on with OpenShift CI, it seems like our test script isn't even getting executed. I've followed up with QE to see what's going on.

**Comment by l0rd** [comment]
@amisevsk I should have address most of your points in the last 4 commits, please have a look.

**Review by amisevsk — APPROVED** [review]
LGTM :+1:

**Pull request merged**
